usr/lib/soft_stdll/soft_specific.c

```diff
@@ -2970,13 +3629,132 @@ CK_RV token_specific_tdes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
     }
 
     return rv;
+#else
+    int rc;
+    size_t maclen;
+    CK_RV rv = CKR_OK;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_KEY_TYPE keytype;
+    const EVP_CIPHER *cipher;
+    struct cmac_ctx {
+        EVP_MD_CTX *mctx;
+        EVP_PKEY_CTX *pctx;
+        EVP_PKEY *pkey;
+    };
+    struct cmac_ctx *cmac = NULL;
+
+    UNUSED(tokdata);
+
+    if (first) {
+        // get the key type
+        rc = template_attribute_find(key->template, CKA_KEY_TYPE, &attr);
+        if (rc == FALSE) {
+            TRACE_ERROR("Could not find CKA_KEY_TYPE for the key.\n");
+            rv = CKR_FUNCTION_FAILED;
+            goto err;
+        }
+        keytype = *(CK_KEY_TYPE *) attr->pValue;
+
+        // get the key value
+        if (template_attribute_find(key->template, CKA_VALUE, &attr) == FALSE) {
+            TRACE_ERROR("Could not find CKA_VALUE for the key.\n");
+            rv = CKR_FUNCTION_FAILED;
+            goto err;
+        }
+        switch (keytype) {
+        case CKK_DES2:
+            cipher = EVP_des_ede_cbc();
+            break;
+        case CKK_DES3:
+            cipher = EVP_des_ede3_cbc();
+            break;
+        default:
+            TRACE_ERROR("Invalid key type: %lu\n", keytype);
+            rv = CKR_KEY_TYPE_INCONSISTENT;
+            goto err;
+        }
+
+        cmac = calloc(1, sizeof(*cmac));
+        if (cmac == NULL) {
+            TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+            rv = ERR_HOST_MEMORY;
+            goto err;
+        }
+
+        cmac->mctx = EVP_MD_CTX_new();
+        if (cmac->mctx == NULL) {
+            TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+            rv = ERR_HOST_MEMORY;
+            goto err;
+        }
+
+        cmac->pkey = EVP_PKEY_new_CMAC_key(NULL,
```

Comment:
I get compile errors when compiling on my system:

```
usr/lib/soft_stdll/.libs/opencryptoki_stdll_libpkcs11_sw_la-soft_specific.o: In function `token_specific_tdes_cmac':
/root/opencryptoki-pr292/usr/lib/soft_stdll/soft_specific.c:3691: undefined reference to `EVP_PKEY_new_CMAC_key'
usr/lib/soft_stdll/.libs/opencryptoki_stdll_libpkcs11_sw_la-soft_specific.o: In function `token_specific_aes_cmac':
/root/opencryptoki-pr292/usr/lib/soft_stdll/soft_specific.c:3883: undefined reference to `EVP_PKEY_new_CMAC_key'
```
with this OpenSSL version:
```
# openssl version
OpenSSL 1.1.0h-fips  27 Mar 2018
```
I guess EVP_PKEY_new_CMAC_key only exists since OpenSSL 1.1.1.
So I guess the OpenSSL version check in this function is not correct:
`#if OPENSSL_VERSION_NUMBER < 0x10100000L`

Reply:
yes, should check for 1.1.1

Reply:
done